### PR TITLE
ci(*): update publish/deploy to use ghcr.io; untrack static/data.json

### PR DESCRIPTION
**Pull request by vdice**
Updates publish/deploy logic to use an OCI registry (currently: docker.io). In doing so, we generate the search index (static/data.json) at time of publishing, removing the need to track in version control. (Ref https://github.com/fermyon/developer/issues/369).

The first choice was ghcr.io, however, in my testing, the GH action runner nearly always encountered rate limits when attempting to push this developer website app (for example: https://github.com/vdice/developer/actions/runs/4892021649/jobs/8733252133).  Follow-up created here: https://github.com/fermyon/developer/issues/589

Successful deploy to https://canary.developer.fermyon.com in https://github.com/fermyon/developer/actions/runs/4892766363/jobs/8734931145.

You can also run the website locally from a sample tag pushed from this branch, eg `spin up -f fermyon/developer:vdice-oci-deploy-d17b043d0884dab9fc2197709cd3035bd4fe4667`

TODO:
- [x] reviews/feedback
- [x] verify search index is functioning as expected (@karthik2804 :smile:)
- [x] address remaining TODOs in deploy.yml prior to merge

Supersedes https://github.com/fermyon/developer/pull/423

Closes https://github.com/fermyon/developer/issues/369

## Pre-Merge Content Checklist

This documentation has been checked to ensure that:

- [ ] The `title, `template`, and `date` are all set
- [ ] Does this PR have a new menu item (anywhere in `templates/*.hbs` files) that points to a document `.md` that is set to publish in the future? If so please only publish the `.md` and `.hbs` changes in real-time (otherwise there will be a menu item pointing to a `.md` file that does not exist)
- [ ] File does not use CRLF, but uses plain LF (hint: use `cat -ve <filename> | grep '^M' | wc -l` and expect 0 as a result) 
- [ ] Has passed [`bart check`](https://developer.fermyon.com/bartholomew/quickstart)
- [ ] Has run `npm run build-index`
- [ ] Has been manually tested by running in Spin/Bartholomew (hint: use `PREVIEW_MODE=1` and run `npm run styles` to update styling)
- [ ] Headings are using Title Case
- [ ] Code blocks have the programming language set to properly highlight syntax and the proper copy directive
- [ ] Have tested with `npm run test` and resolved all errors

### CODE REVIEW

**Comment by vdice** [comment]
One item to note on the Nomad side.  With the updated website job (now running `spin up -f <oci ref>`), CPU spikes quite high when starting up, I assume from the high number of concurrent downloads when pulling the app/image.  Once running, CPU usage drops back down to the usual ~10 MHz idle baseline.

Screenshot 2023-05-05 at 1 18 23 PM

**Comment by vdice** [comment]
Note: Build failure currently tracked in https://github.com/fermyon/developer/issues/590

File: .github/workflows/deploy-preview.yml [hunk 1]
```diff
@@ -17,6 +17,16 @@ jobs:
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
 
+      - name: Install npm packages
+        run: |
+          npm ci
+
+      # TODO: any other node-y things to do prior to build/deploy? 'npm run styles', etc.?
```
Comment (karthik2804) [inline .github/workflows/deploy-preview.yml:24]: We could run `npm run styles` but have so far been pushing up the CSS files with PRs. We could add it if we want to stop tracking the generated CSS files as well. @flynnduism any preferences here?

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -38,12 +40,73 @@ jobs:
...
+            registry: ${{ env.OCI_REGISTRY }}
+            username: fermyon
+            password: ${{ secrets.DOCKERHUB_PAT }}
+
+        - name: Publish Spin app - mutable tag
+          if: ${{ github.ref == 'refs/heads/main' }}
+          run: |
+            spin registry push ${{ env.OCI_REGISTRY }}/${{ env.OCI_IMAGE }}:latest
+
+        - name: Publish Spin app - immutable tag
```
Comment (karthik2804) [inline .github/workflows/deploy.yml:97]: Any reason we want to keep an immutable tag around?

**Comment by karthik2804** [comment]
This looks great @vdice. The search index seems to be working as expected. I will wait to approve until we get all the todos cleaned and fixed but it looks good to me!

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -38,12 +40,73 @@ jobs:
...
+          uses: docker/login-action@v2.1.0
+          with:
+            registry: ${{ env.OCI_REGISTRY }}
+            username: fermyon
+            password: ${{ secrets.DOCKERHUB_PAT }}
+
+        - name: Publish Spin app - mutable tag
+          if: ${{ github.ref == 'refs/heads/main' }}
+          run: |
+            spin registry push ${{ env.OCI_REGISTRY }}/${{ env.OCI_IMAGE }}:latest
```
Comment (rajatjindal) [inline .github/workflows/deploy.yml:95]: there is a Github action for this. (to build/push to OCI). we should give that a shot.

Comment (vdice) [reply .github/workflows/deploy.yml:95]: Oh right, thank you for the reminder @rajatjindal.  I'll update.

File: .github/workflows/deploy-preview.yml (hunk 1 above)
Comment (vdice) [reply .github/workflows/deploy-preview.yml:24]: Created https://github.com/fermyon/developer/issues/609 as a potential follow-up.

**Comment by vdice** [comment]
This is ready for a final review @karthik2804 @rajatjindal.  Also note https://github.com/fermyon/developer/pull/588#discussion_r1187518287.  Thanks!

**Review by tpmccallum — APPROVED** [review]
This looks good to me @vdice

**Pull request merged**